Tasks/Maven/maventask.ts

```diff
@@ -170,11 +186,121 @@ function publishJUnitTestResults(testResultsFiles: string) {
+    var buildRootPath = path.dirname(mavenPOMFile);
+    // appending with small guid to keep it unique. Avoiding full guid to ensure no long path issues.
+    var reportPOMFileName = "CCReportPomA4D283EG.xml";
+    reportPOMFile = path.join(buildRootPath, reportPOMFileName);
+    var targetDirectory = path.join(buildRootPath, "target");
+    ccReportTask = "jacoco:report";
+
+    if (ccTool.toLowerCase() == "jacoco") {
+        var reportDirectoryName = "CCReport43F6D5EF";
+        var summaryFileName = "jacoco.xml";
+    }
+    else if (ccTool.toLowerCase() == "cobertura") {
+        var reportDirectoryName = "target/site/cobertura";
+        var summaryFileName = "coverage.xml";
+    }
+
+    reportDirectory = path.join(buildRootPath, reportDirectoryName);
+    summaryFile = path.join(reportDirectory, summaryFileName);
+
+    if (ccTool.toLowerCase() == "jacoco") {
+        execFileJacoco = path.join(reportDirectory, "jacoco.exec");
+    }    
+        
+    // clean any previously generated files.
+    tl.rmRF(targetDirectory, true);
+    tl.rmRF(reportDirectory, true);
+    tl.rmRF(reportPOMFile, true);
+
+    var buildProps: { [key: string]: string } = {};
+    buildProps['buildfile'] = mavenPOMFile;
+    buildProps['classfilter'] = classFilter
+    buildProps['classfilesdirectories'] = classFilesDirectories;
+    buildProps['sourcedirectories'] = sourceDirectories;
+    buildProps['summaryfile'] = summaryFile;
+    buildProps['reportdirectory'] = reportDirectory;
+    buildProps['reportbuildfile'] = reportPOMFile;
+
+    try {
+        var codeCoverageEnabler = new tl.CodeCoverageEnabler('Maven', ccTool);
+        codeCoverageEnabler.enableCodeCoverage(buildProps);
+        tl.debug("Code coverage is successfully enabled.");
+    }
+    catch (Error) {
+        tl.warning("Enabling code coverage failed. Check the build logs for errors.");
+    }
+}
+
+function publishCodeCoverage(isCodeCoverageOpted: boolean) {
+    if (isCodeCoverageOpted) {
+        tl.debug("Collecting code coverage reports");
+
+        if (ccTool.toLowerCase() == "jacoco") {
+            var mvnReport = tl.createToolRunner(mvnExec);
+            mvnReport.arg('-f');
+            if (isFileExists(reportPOMFile)) {
+                // multi module project
+                mvnReport.pathArg(reportPOMFile);
+                mvnReport.arg("verify");
+            }
+            else {
+                mvnReport.pathArg(mavenPOMFile);
+                mvnReport.arg(ccReportTask);
+            }
+            mvnReport.exec().then(function(code) {
+                publishCCToTfs();
+            }).fail(function(err) {
+                tl.warning("No code coverage found to publish. There might be a build failure resulting in no code coverage or there might be no tests.");
+            });
+        }
+        else if (ccTool.toLowerCase() == "cobertura") {
+            publishCCToTfs();
+        }
+    }
+}
+
+function publishCCToTfs() {
```

Comment:
It is confusing to have 2 functions named so similarly: publishCodeCoverage() and publishCCToTfs().  What if you did something like this?
```

function publishCodeCoverage(isCodeCoverageOpted: boolean) {
    if (isCodeCoverageOpted) {
        tl.debug("Collecting code coverage reports");
        if (********collectCodeCoverageReports()********) {
            if (isFileExists(summaryFile)) {
                tl.debug("Summary file = " + summaryFile);
                tl.debug("Report directory = " + reportDirectory);
                tl.debug("Publishing code coverage results to TFS");
                var ccPublisher = new tl.CodeCoveragePublisher();
                ccPublisher.publish(ccTool, summaryFile, reportDirectory, "");
            }
            else {
                tl.warning("No code coverage found to publish. There might be a build failure resulting in no code coverage or there might be no tests.");
            }
        }
    }
}
```